manifests/smartctl_exporter.pp

```diff
@@ -0,0 +1,251 @@
+  Boolean $purge_config_dir                                  = true,
+  Boolean $restart_on_change                                 = true,
+  Boolean $service_enable                                    = true,
+  Stdlib::Ensure::Service $service_ensure                    = 'running',
+  String[1] $service_name                                    = 'smartctl_exporter',
+  Prometheus::Initstyle $init_style                          = $prometheus::init_style,
+  Prometheus::Install $install_method                        = $prometheus::install_method,
+  Boolean $manage_group                                      = false,
+  Boolean $manage_service                                    = true,
+  Boolean $manage_user                                       = false,
+  String[1] $os                                              = downcase($facts['kernel']),
+  String $options                                            = '', # lint:ignore:params_empty_string_assignment
+  Optional[Prometheus::Uri] $download_url                    = undef,
+  String[1] $arch                                            = $prometheus::real_arch,
+  Stdlib::Absolutepath $bin_dir                              = $prometheus::bin_dir,
+  Boolean $export_scrape_job                                 = false,
+  Optional[Stdlib::Host] $scrape_host                        = undef,
+  Stdlib::Port $scrape_port                                  = 9633,
+  String[1] $scrape_job_name                                 = 'smartctl',
+  Optional[Hash] $scrape_job_labels                          = undef,
+  Stdlib::Absolutepath $web_config_file                      = '/etc/smartctl_exporter_web-config.yml',
+  Optional[Hash] $web_config_content                         = undef,
+  Optional[String[1]] $proxy_server                          = undef,
+  Optional[Enum['none', 'http', 'https', 'ftp']] $proxy_type = undef,
+) inherits prometheus {
+  $release = "v${version}"
+
+  $real_download_url = pick($download_url, "${download_url_base}/download/${release}/${package_name}-${version}.${os}-${arch}.${download_extension}") # lint:ignore:140chars
+
+  $notify_service = $restart_on_change ? {
+    true    => Service[$service_name],
+    default => undef,
+  }
+
+  $_web_config_ensure = $web_config_content.empty ? {
+    true    => absent,
+    default => file,
+  }
+
+  file { $web_config_file:
+    ensure  => $_web_config_ensure,
+    owner   => $user,
+    group   => $group,
+    mode    => '0640',
+    content => $web_config_content.stdlib::to_yaml,
+    notify  => $notify_service,
+  }
+
+  $_web_config = if $web_config_content.empty {
+    ''
+  } else {
+    "--web.config.file=${$web_config_file}"
+  }
+
+  $_cmd_smartctl_device = $smartctl_device.map |$device| {
+    "--smartctl.device=${device}"
+  }
+
+  $_cmd_web_listen_address = $web_listen_address.map |$address| {
+    "--web.listen-address=${address}"
+  }
+
+  $_smartctl_path = $smartctl_path ? {
```

Comment:
Why set the defaults for all these parameters to undef and then override them here. Instead remove the `Optional` and set the default value directly in the parameter definition.

Reply:
That's not quite what's happening here. The intent here was to support [smartctl's optional parameters](https://github.com/prometheus-community/smartctl_exporter?tab=readme-ov-file#command-line-options) that, when passed, require them to be passed with specific named flags.

To be clear about the section you highlighted: 

If the user sets a smartctl_path, then pass it using `--smartctl.path=`
If the user does not set a smartctl_path, then don't pass anything.

In other words, smartctl_path remains an optional flag. If I removed `Optional` and set the default value in the parameter definition per your suggestion, I would be sending invalid (empty) flags to the exporter.

I tried to follow the pattern set in other exporters in this project. See e.g. this: 
https://github.com/voxpupuli/puppet-prometheus/blob/44e45dd1639d0e516a7becadbf4168768eb88f9d/manifests/openvpn_exporter.pp#L95-L98

If you have another preferred pattern to use in instances like this, let me know.

Reply:
right, i overlooked that.

The usual pattern is to just pass in `extra_options` just search for it in the code multiple example

That being said, I already started a discussion about moving all the exporters out of this module. I am quite reluctant to to the ever growing list of exporters in this module.

So if you want to have exporter specific parameters, maybe a dedicated module is a better option. It could still be depending on puppet-prometheus and re-use all of the shared bits.
But I that's just my opinion at the moment